kafkaesk/consumer.py

```diff
@@ -0,0 +1,400 @@
+
+    async def _consumer_factory(self) -> aiokafka.AIOKafkaConsumer:
+        consumer = self._app.consumer_factory(self.group_id)
+        # This is needed in case we have a prefix
+        pattern = fnmatch.translate(self._app.topic_mng.get_topic_id(self.stream_id))
+        consumer.subscribe(pattern=pattern, listener=self)
+        return consumer
+
+    async def stop(self) -> None:
+        if not self._running:
+            return
+
+        # Exit the loop, this will trigger finalize call
+        loop = asyncio.get_running_loop()
+        self._close = loop.create_future()
+        await asyncio.wait([self._close])
+
+    def __repr__(self) -> str:
+        return f"<Consumer: {self.stream_id}, Group: {self.group_id}>"
+
+    def _span(self, record: aiokafka.ConsumerRecord) -> opentracing.SpanContext:
+        tracer = opentracing.tracer
+        headers = {x[0]: x[1].decode() for x in record.headers or []}
+        parent = tracer.extract(opentracing.Format.TEXT_MAP, headers)
+        context = tracer.start_active_span(
+            record.topic,
+            tags={
+                "message_bus.destination": record.topic,
+                "message_bus.partition": record.partition,
+                "message_bus.group_id": self.group_id,
+            },
+            references=[opentracing.follows_from(parent)],
+        )
+        return context.span
+
+    async def _handler(self, record: aiokafka.ConsumerRecord) -> None:
+        with self._span(record) as span:
+            await self._message_handler(record, span)
+
+    async def _consume(self) -> None:
+        batch = await self._consumer.getmany(max_records=self._concurrency, timeout_ms=500)
+
+        if not batch:
+            await self._maybe_commit()
+            return
+
+        futures: typing.Dict[asyncio.Future[typing.Any], aiokafka.ConsumerRecord] = dict()
+        await self._processing.acquire()
+
+        for tp, records in batch.items():
+            for record in records:
+                coro = self._handler(record)
+                fut = asyncio.create_task(coro)
+                futures[fut] = record
+
+        # TODO: this metric is kept for backwards-compatibility, but should be revisited
+        with CONSUMED_MESSAGE_TIME.labels(
+            stream_id=self.stream_id,
+            partition=next(iter(batch)),
+            group_id=self.group_id,
+        ).time():
+            done, pending = await asyncio.wait(
+                futures.keys(), timeout=self._timeout, return_when=asyncio.FIRST_EXCEPTION
+            )
+
+        # Look for failures
+        for task in done:
+            record = futures[task]
+
+            try:
+                if exc := task.exception():
+                    self._count_message(record, error=exc.__class__.__name__)
+                    await self.on_handler_failed(exc, record)
+                else:
+                    self._count_message(record)
+            except asyncio.CancelledError:
```

Comment:
Dunno but it might be good to have some more explicit code comment to explain that calling `exception` will raise this.